dttools/src/jx.c

```diff
@@ -0,0 +1,410 @@
+/*
+Copyright (C) 2015- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include "jx.h"
+#include "stringtools.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdarg.h>
+
+struct jx_pair * jx_pair( struct jx *key, struct jx *value, struct jx_pair *next )
+{
+	struct jx_pair *pair = malloc(sizeof(*pair));
+	pair->key = key;
+	pair->value = value;
+	pair->next = next;
+	return pair;
+}
+
+struct jx_item * jx_item( struct jx *value, struct jx_item *next )
+{
+	struct jx_item *item = malloc(sizeof(*item));
+	item->value = value;
+	item->next = next;
+	return item;
+}
+
+static struct jx * jx_create( jx_type_t type )
+{
+	struct jx *j = malloc(sizeof(*j));
+	j->type = type;
+	return j;
+}
+
+struct jx * jx_null()
+{
+	return jx_create(JX_NULL);
+}
+
+struct jx * jx_symbol( const char *symbol_name )
+{
+	struct jx *j = jx_create(JX_SYMBOL);
+	j->symbol_name = strdup(symbol_name);
+	return j;
+}
+
+struct jx * jx_string( const char *string_value )
+{
+	struct jx *j = jx_create(JX_STRING);
+	j->string_value = strdup(string_value);
+	return j;
+}
+
+struct jx * jx_string_nodup( char *string_value )
+{
+	struct jx *j = jx_create(JX_STRING);
+	j->string_value = string_value;
+	return j;
+}
+
+struct jx * jx_integer( jx_int_t integer_value )
+{
+	struct jx *j = jx_create(JX_INTEGER);
+	j->integer_value = integer_value;
+	return j;
+}
+
+struct jx * jx_double( double double_value )
+{
+	struct jx *j = jx_create(JX_DOUBLE);
+	j->double_value = double_value;
+	return j;
+}
+
+struct jx * jx_boolean( int boolean_value )
+{
+	struct jx *j = jx_create(JX_BOOLEAN);
+	j->boolean_value = boolean_value;
+	return j;
+}
+
+struct jx * jx_object( struct jx_pair *pairs )
+{
+	struct jx *j = jx_create(JX_OBJECT);
+	j->pairs = pairs;
+	return j;
+}
+
+struct jx * jx_array( struct jx_item *items )
+{
+	struct jx *j = jx_create(JX_ARRAY);
+	j->items = items;
+	return j;
+}
+
+struct jx * jx_arrayv( struct jx *value, ... )
+{
+	va_list args;
+	va_start(args,value);
+
+	struct jx *array = jx_array(0);
+	while(value) {
+		jx_array_append(array,value);
+		value = va_arg(args,struct jx *);
+	}
+
+	return array;
+}
+
+struct jx * jx_lookup( struct jx *j, const char *key )
+{
+	struct jx_pair *p;
+
+	if(!j || j->type!=JX_OBJECT) return 0;
+
+	for(p=j->pairs;p;p=p->next) {
+		if(p && p->key && p->key->type==JX_STRING) {
+			if(!strcmp(p->key->string_value,key)) {
+				return p->value;
+			}
+		}
+	}
+
+	return 0;
+}
+
+const char * jx_lookup_string( struct jx *object, const char *key )
+{
+	struct jx *j = jx_lookup(object,key);
+	if(j && jx_istype(j,JX_STRING)) {
+		return j->string_value;
+	} else {
+		return 0;
+	}
+}
+
+jx_int_t jx_lookup_integer( struct jx *object, const char *key )
+{
+	struct jx *j = jx_lookup(object,key);
+	if(j && jx_istype(j,JX_INTEGER)) {
+		return j->integer_value;
+	} else {
+		return 0;
```

Comment:
Keep in mind that `jx_lookup_integer` is already a common-case wrapper around `jx_lookup` for brevity.  If you care about the alternate cases, you can do this:

```
j = jx_lookup(j,key);
if(!j) { //no such key } 
else if(!jx_istype(JX_STRING) { // unexpected type }
else { // expected case }
```